crypto/ec/curve448/eddsa.c

```diff
@@ -0,0 +1,359 @@
+/*
+ * Copyright 2017-2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2015-2016 Cryptography Research, Inc.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ *
+ * Originally written by Mike Hamburg
+ */
+#include <openssl/crypto.h>
+#include <openssl/evp.h>
+
+#include "curve448_lcl.h"
+#include "word.h"
+#include "ed448.h"
+#include <string.h>
+#include "internal/numbers.h"
+
+#define COFACTOR 4
+
+static c448_error_t oneshot_hash(uint8_t *out, size_t outlen,
+                                 const uint8_t *in, size_t inlen)
+{
+    EVP_MD_CTX *hashctx = EVP_MD_CTX_new();
+
+    if (hashctx == NULL)
+        return C448_FAILURE;
+
+    if (!EVP_DigestInit_ex(hashctx, EVP_shake256(), NULL)
+            || !EVP_DigestUpdate(hashctx, in, inlen)
+            || !EVP_DigestFinalXOF(hashctx, out, outlen)) {
+        EVP_MD_CTX_free(hashctx);
+        return C448_FAILURE;
+    }
+
+    EVP_MD_CTX_free(hashctx);
+    return C448_SUCCESS;
+}
+
+static void clamp(uint8_t secret_scalar_ser[EDDSA_448_PRIVATE_BYTES])
+{
+    uint8_t hibit = (1 << 0) >> 1;
```

Comment:
hey, what is this `(1 << 0) >> 1` ?

Reply:
It's a remnant of the original code approach. We've just imported the curve448 stuff, but the original codebase supported 25519 as well. Some of the code was templated and reused between the two implementations. I removed all of that as part of this work. This code originally looked like this:

````
    uint8_t hibit = (1<<$(gf_bits % 8))>>1;
````
I have now simplified this function accordingly.